certora/specs/sharesAssetsConversion.spec

```diff
@@ -0,0 +1,257 @@
+ * - shares to assets = A(s) = (s * r) // R
+ * where a=assets, s=shares, R=RAY, r=rate
+ * 
+ * These imply:
+ *   a * R - r <= S(a) * r <= a * R    a*R/r - 1 <= S(a) <= a*R/r
+ *   s * r - R <= A(s) * R <= s * r    s*r/R - 1 <= A(s) <= s*r/R
+ * 
+ * Hence:
+ *   A(S(a)) >= S(a)*r/R - 1 >= (a*R/r - 1)*r/R - 1 = (a*R - r)/R - 1 = a - r/R - 1
+ *   S(A(s)) >= A(s)*R/r - 1 >= (s*r/R - 1)*R/r - 1 = (s*r - R)/r - 1 = s - R/r - 1
+ */
+
+
+// Converting amount to shares is rounded down.
+rule amountConversionRoundedDown(uint256 amount) {
+	uint256 shares = convertToShares(amount);
+	assert convertToAssets(shares) <= amount, "Too many converted shares";
+	assert convertToAssets(shares + 1) >= amount, "Too few converted shares";
+}
+
+
+// Converting shares to amount is rounded down.
+rule sharesConversionRoundedDown(uint256 shares) {
+	uint256 amount = convertToAssets(shares);
+	assert convertToShares(amount) <= shares, "Amount converted is too high";
+	assert convertToShares(amount + 1) >= shares, "Amount converted is too low";
+}
+
+
+// Converting amount to shares and back to amount is preserved (up to rounding).
+/* NOTE: The precision depends on the ratio rate/ray.
+ */
+rule amountConversionPreserved(uint256 amount) {
+ 	require rate() <= RAY();
+ 	mathint mathamount = to_mathint(amount);
+ 	mathint converted = to_mathint(convertToAssets(convertToShares(amount)));
+
+	// That converted <= mathamount was proved in amountConversionRoundedDown,
+ 	assert mathamount - converted <= 1 + rate() / RAY(), "Too few converted assets";
+ }
+ 
+
+// Converting shares to amount and back to shares is preserved up to RAY.
+/* NOTE: The precision depends on the ratio RAY/rate.
+ */
+rule sharesConversionPreserved(uint256 shares) {
+	mathint mathshares = to_mathint(shares);
+	uint256 amount = convertToAssets(shares);
+	mathint converted = to_mathint(convertToShares(amount));
+
+	// That converted <= mathshare was proved in sharesConversionRoundedDown.
+	assert mathshares - converted <= 1 + RAY() / rate(), "Too few converted shares";
+}
+
+
+/* Joining and splitting accounts provides limited advantage.
+ * This rule verifies that joining accounts (by combining shares), and splitting accounts
+ * (by splitting shares between accounts) provides limited advantage when converting to
+ * asset amounts.
+ */
+rule accountsJoiningSplittingIsLimited(uint256 shares1, uint256 shares2) {
+    uint256 amount1 = convertToAssets(shares1);
+    uint256 amount2 = convertToAssets(shares2);
+    uint256 jointShares = shares1 + shares2;
+    require jointShares >= shares1 + shares2;  // Prevent overflow
+    uint256 jointAmount = convertToAssets(jointShares);
+
+    assert jointAmount >= amount1 + amount2, "Found advantage in combining accounts";
+    assert jointAmount < amount1 + amount2 + 2, "Found advantage in splitting accounts";
+    // The following assertion fails (as expected):
+    // assert jointAmount < amount1 + amount2 + 1, "Found advantage in splitting accounts";
```

Comment:
like the rule below, here it seems more reasonable, but lets discuss the reason joining can gain more and decide if it's ok. Maybe you already have the answer, im still not clear on that

Reply:
I added a comment explaining this.